clang/include/clang/Serialization/ASTReader.h

```diff
@@ -1341,9 +1341,22 @@ class ASTReader
   serialization::InputFile getInputFile(ModuleFile &F, unsigned ID,
                                         bool Complain = true);
 
+  /// Buffer we use as temporary storage backing resolved paths.
+  SmallString<256> PathBuf;
```

Comment:
Such global buffer is easy to get wrong... I fear it might introduce maintain burden to every one.

Reply:
Do you have any suggestions?

Reply:
How bad will it be to ask the callers to provide the buffer (SmallString) all time?

Reply:
I'd be fine with explicitly passing the global buffer into `ResolveImportedPath()` at all call-sites, and I can see how it clarifies the intent.

Reply:
I am wondering why we need the buffer to be global? What will happen if we ask the caller to own the buffer?

Reply:
Because if callers construct fresh buffers all the time, each call to `ResolveImportedPath()` allocates. If the buffer is global, the function never allocates again after the longest path has been resolved.

Reply:
Sounds... not too bad?

If we really want it, can we make the so-called global buffer to be a static local variable?

Reply:
Allocations during path reading account for 5+% of wall time of the entire dependency scan, so I have to disagree that this isn't so bad. I'm curious, how is static local better than member variable?

Reply:
I think the static local variable makes it harder to be misused. As you said, we must be careful for the "global" variable. Then I feel better if we can reduce the mind burden.

Reply:
I don't see how static local changes anything. If you don't get the lifetimes right, you get bogus results anyway. It just changes when and what will
overwrite the underlying buffer.

Reply:
@ChuanqiXu9 a static local would be global to the process instead of the compilation, which is unsafe if you have two independent `CompilerInstance` running in parallel.

@jansvoboda11 just an idea: would it help if we change the return type to some RAII object that takes temporary ownership of the path buffer? It would prevent the new kind of misuse: multiple calls to `ResolveImportedPath` with overlapping lifetimes for their results. It wouldn't prevent escaping a StringRef in callers (e.g. when calling `visitInputFile`), but that's also true today with `std::string`.

Reply:
@benlangmuir I tried sketching this in the latest commit. Let me know what you think.

Reply:
Note that I don't wrap the return value in `TemporarilyOwnedStringRef` for the static versions of `ResolveImportedPath()`. Those take the buffer as out-parameter, which makes it fairly obvious (along with the documentation) how the lifetimes work. I'd be in favor of just explicitly passing the member `ASTReader` buffer into the same function and call it obvious enough. But LMK if you'd like to go the other direction and extend `TemporarilyOwnedStringRef` to the static functions too.